### PR TITLE
DEV: Update development Docker image 

### PR DESCRIPTION
**Pull request by trallard**
#### Reference issue

NA

#### What does this implement/fix?

This PR adds the following:
-  Adds a new Dockerfile with all the compilers and dependencies needed to set a SciPy development environment
- Adds a github actions workflow to automate the building, tagging and pushing of Docker images (also adds linting so added the relevant `.hadolint.yml` file too)
-  Updates the [Docker quickstart instructions](https://docs.scipy.org/doc/scipy/reference/dev/contributor/quickstart_docker.html#quickstart-docker) to reflect the changes introduced by the new Docker image

#### Additional information

TODO:
- Get DockerHub token and add as a secret for the GH actions workflow

Also at the moment, this image is tagged as `scipy/scipy-dev:date-branch-sha8`  since it would only need rebuilding when updating dependencies or the such. 

Note there is a `Dockerfile` in `tools/`, but this PR would effectively replace it - I left it in case it was needed for review purposes or else.

Finally, this Docker image will be later used as the `base` for a `gitpod` development environment

### CODE REVIEW

**Comment by ilayn** [comment]
Just for my own understanding, do you want SciPy to provide official docker images?

**Comment by trallard** [comment]
Hey @ilayn  yes, SciPy would provide the official Docker images 
An added benefit is that these will later be used to enable the use of <https://gitpod.io/> 

I also forgot to cc/ @rgommers  for visibility 👀

**Comment by ev-br** [comment]
So that the scipy team maintains this image, too?

**Comment by trallard** [comment]
I understand that this does drop more stuff on your plate and why you'd be hesitant to accept it. So here I see two clear options 

1. Remove the GH actions, update the Docker file and leave it as is with folks being responsible for building their own image (vanilla Docker) 

2. I can help with the maintenance of both the Docker images and the gitpod as I'm submitting this PR 


Of course there might be another alternative and I'd be happy to discuss

**Comment by ilayn** [comment]
Not necessarily hesitant but I'm not familiar enough with docker maintenance and what it entails (I do use it for work albeit only pulling images from here and there). So I guess more docker-fluent members can chime in for it. I just wanted to clarify for myself.

**Comment by rgommers** [comment]
I'm definitely +1 on this - I have wanted Gitpod integration for a long time, it would be great for new contributors to get started (or even experienced devs, building SciPy locally can be more trouble than it's worth).

Also, @trallard is my colleague and one of our resident DevOps gurus, so I'm not too worried about maintenance load if she says she will help with maintenance:)

> I think the versioning of the images and backporting would be a bit trickier?

Why would we need to backport anything? The dev image only changes when we add dependencies, which can't happen in maintenance branches (and is rare in general). The builds of SciPy itself also shouldn't be special regarding backports (I think).

**Comment by ilayn** [comment]
Wouldn't it be weird that we only provide dev images and not a stable latest scipy image? Then we need some versioning no?

**Comment by trallard** [comment]
The idea behind having this official Docker image and gitpod.io is to help with the contributors' onboarding and setup.  gitpod is especially useful if folks only want to make some minor changes to the docs for example or contributions here and there without the overhead of having a working local environment. 

For all those other folks that use SciPy for work (usually with a bunch of other libraries) , I expect they and their teams handle these on their own for their specific purposes.

**Comment by rgommers** [comment]
I think in general it doesn't make sense to provide a Docker image for end users for any particular Python package. It's trivial to bake your own with exactly the set of packages you want (using wheels from PyPI), and much more flexible.

File: .github/workflows/docker.yml
Comment (rgommers) [inline .github/workflows/docker.yml:30]: Why is QEMU needed? It tends to be slow and not completely reliable, native builds should work just fine here right?

File: .github/workflows/docker.yml
Comment (rgommers) [inline .github/workflows/docker.yml:45]: Probably also needs a condition that the branch is master. There's no need for building on maintenance branches.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [inline tools/Docker-dev/Dockerfile:7]: minor: Directory and file names exclusively use underscores, and all-lowercase would also be easier. Could you change to `docker_dev`.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [inline tools/Docker-dev/Dockerfile:37]: Just a note: I checked the time the solve takes, since that can be problematic in CI. Here it's ~12 seconds, which is fine.

File: tools/Docker-dev/environment.yml
Comment (rgommers) [inline tools/Docker-dev/environment.yml:11]: Could you add `pythran=0.9.8.post2` (it's an optional build dependency)

Also worth considering to add:
```
- ipython=7.19.0  # that version has working autocompletion it looks like
```
it's kind of an essential development tool I think. Or maybe it could come from somewhere else, I don't know much about how best to layer the images.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [inline tools/Docker-dev/Dockerfile:117]: Could you add a comment for this line? I'm not quite sure what it does.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [inline tools/Docker-dev/Dockerfile:87]: Probably better to install `openblas` with conda instead, and remove this.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [inline tools/Docker-dev/Dockerfile:81]: Should not be needed I think, Python development headers must come from conda.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [inline tools/Docker-dev/Dockerfile:92]: These are optional dependencies. I think it makes sense to add them here even though technically it's better to get them from conda-forge if possible. We never touch them and they do help make the tests run faster.

Comment (rgommers) [reply tools/Docker-dev/Dockerfile:92]: It could conceivably lead to problems though, because this is also why `gfortran` is needed here, and it's ignored because there's a `gfortran` in the `compilers` conda package too.

**Review by rgommers — COMMENTED** [review]
Thanks Tania, looks great! A few questions and comments.

**Comment by ilayn** [comment]
> I think in general it doesn't make sense to provide a Docker image for end users for any particular Python package. 

Yes this is why I was surprised that we are providing one now. But I leave it to you folks since I don't have enough experience with it.

File: tools/Docker-dev/Dockerfile
Comment (ilayn) [reply tools/Docker-dev/Dockerfile:87]: Isn't it already coming with MKL out of the box?

**Comment by rgommers** [comment]
> Yes this is why I was surprised that we are providing one now.

This PR doesn't - it's a development tool only, no end user should be using it.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [reply tools/Docker-dev/Dockerfile:87]: Right - the default on conda-forge is OpenBLAS. They have this interesting `blas` metapackage where you can switch to/from Netlib, MKL, BLIS, etc. after numpy is already built. I think openblas indeed gets pulled in implicitly.

File: .github/workflows/docker.yml
Comment (trallard) [reply .github/workflows/docker.yml:30]: Ah yes - we could use just native builds (carried this over from other action I have in another repo, will remove it)

File: .github/workflows/docker.yml
Comment (trallard) [reply .github/workflows/docker.yml:45]: At the moment this action should only run on pushes to master and new tags. From above:

```
on:
  push: 
    branches:
      - master
    tags:
      - '*'
   ```

We can add other conditions as needed

File: tools/Docker-dev/Dockerfile
Comment (trallard) [reply tools/Docker-dev/Dockerfile:92]: Right - I was nor sure which deps are **mandatory** as I looked at the Dockerfile already existing in the repo.

But I can remove anything that is not necessary and ensure conda installs superseeds these

File: tools/Docker-dev/Dockerfile
Comment (trallard) [reply tools/Docker-dev/Dockerfile:117]: Will do - this allows for the Python version (currently 3.8) to be changed at build-time by providing a build arg

File: tools/Docker-dev/environment.yml
Comment (trallard) [reply tools/Docker-dev/environment.yml:11]: Also - would it be worth adding `conda-build`? I noticed the Linux instructions with Conda do recommend using `conda develop scipy` and I could perhaps use this as an initialization command later on for gitpod

Comment (rgommers) [reply tools/Docker-dev/environment.yml:11]: Yes that makes sense to add.

File: tools/Docker-dev/Dockerfile
Comment (rgommers) [reply tools/Docker-dev/Dockerfile:92]: The [conda-forge recipe](https://github.com/conda-forge/scipy-feedstock/blob/master/recipe/meta.yaml) doesn't have any of these. They're not necessary to build, and they don't speed up the tests either unless `mpmath` is installed. So leaving all of this stuff out for now is fine I think. Maybe just add a comment in `environment.yml`:
```
# Optional dependencies left out: mpmath, gmpy2, libsuitesparse, libgmp, libmpfr, libsuitesparse, libmpc, sparse, scikit-umfpack
```

File: tools/Docker-dev/Dockerfile
Comment (trallard) [reply tools/Docker-dev/Dockerfile:87]: So I removed this (and got openblas from conda) and got the dreaded:

```
 ImportError: libf77blas.so.3: cannot open shared object file: No such file or directory
```

which resolves installing `libatlas-base-dev`

**Review by rgommers — APPROVED** [review]
This LGTM now. We can only test the GH action after merge, but I assume it works once we put in `DOCKERHUB_USERNAME` and `DOCKERHUB_TOKEN`. Should I do this @trallard? I haven't done that before on DockerHub, any advice on how to best do this?

**Comment by trallard** [comment]
Yes, that would be good @rgommers. The steps are as follow:

1. Head to <https://hub.docker.com/settings/security> 
2. Click on the `New Access Token` button 
3.  Provide a description and click on `Create` as any other PAT you'll need to copy the token as it's only displayed on creation
4. Head over to <https://github.com/scipy/scipy/settings/environments> click on `Create new environment` button
5. Create a new env `scipy-dev` (as it's the env the gh-action is using) and click on `Configure environment`
6.  In the `Environment secrets` create `DOCKERHUB_USERNAME` and `DOCKERHUB_TOKEN` (you'll have to use your DockerHub username as there is no way to login to an org atm)

Done 🎉

**Comment by rgommers** [comment]
> The steps are as follow:

Thanks, done!

**Comment by rgommers** [comment]
In it goes. Thanks @trallard!

**Comment by rgommers** [comment]
This PR missed the code to ensure the action doesn't get run on forks, as in gh-12921. I can try to submit a fix later, or maybe you want to @trallard?

**Comment by trallard** [comment]
Will send a patch later today 😉

**Comment by rlucas7** [comment]
@trallard 
I just went through the updated instructions and they work well-thanks for these.
It does look like the image ***does not*** have git client installed in the container.  
The `apt` listing is cleaned so I needed to do: 

```
apt-get update
apt install git
```

inside the container. I looked at [this post]([url](https://stackoverflow.com/questions/29929534/docker-error-unable-to-locate-package-git)) and IIUC adding this line: 

```
RUN apt-get update && apt-get install -y git
```
in the DOCKER file would have git installed by default. 

Do we want `git` installed by default in the container too? 

If yes, I can open a PR to add.

**Pull request merged**
